### PR TITLE
feat: resource tags in dataset

### LINKED ISSUE #2091: Attach resource tags on datasets
Thanks for stopping by to let us know something could be better!

**PLEASE READ**: If you have a support contract with Google, please create an issue in the [support console](https://cloud.google.com/support/) instead of filing on GitHub. This will ensure a timely response.

 **Is your feature request related to a problem? Please describe.**

We cannot attach [resource tags](https://cloud.google.com/bigquery/docs/tags#tag_datasets) on datasets via Python library. Only possible via REST api.

 **Describe the solution you'd like**

Being able to attach resource tags in Datasets.

 **Describe alternatives you've considered**

 **Additional context**

### PR DESCRIPTION
**Pull request by keunsoopark**
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes https://github.com/googleapis/python-bigquery/issues/2091 🦕

### CODE REVIEW

File: google/cloud/bigquery/dataset.py [hunk 1]
```diff
@@ -801,6 +802,26 @@ def labels(self, value):
...
         self._properties["labels"] = value
 
+    @property
+    def resource_tags(self):
+        """Dict[str, str]: Resource tags of the dataset.
+
+        This method always returns a dict. To change a dataset's resource tags,
+        modify the dict, then call
+        :meth:`google.cloud.bigquery.client.Client.update_dataset`. To delete
+        a resource tag, set its value to :data:`None` before updating.
```
Comment (Linchin) [inline google/cloud/bigquery/dataset.py:812]: Change the docstring to be consistent with [the API doc](https://cloud.google.com/bigquery/docs/reference/rest/v2/datasets).
```suggestion
        Optional. The tags attached to this dataset. Tag keys are globally
        unique. Tag key is expected to be in the namespaced format, for
        example "123456789012/environment" where 123456789012 is
        the ID of the parent organization or project resource for this tag
        key. Tag value is expected to be the short name, for example
        "Production".
```

File: google/cloud/bigquery/dataset.py [hunk 2]
```diff
@@ -801,6 +802,26 @@ def labels(self, value):
...
+
+        Raises:
+            ValueError: for invalid value types.
+        """
+        return self._properties.setdefault("resourceTags", {})
+
+    @resource_tags.setter
+    def resource_tags(self, value):
+        if not isinstance(value, dict):
```
Comment (Linchin) [inline google/cloud/bigquery/dataset.py:821]:
```suggestion
        if not isinstance(value, dict) or value is not None:
```

**Review by Linchin — CHANGES_REQUESTED** [review]
Thank you so much for adding resource tags in dataset! The PR looks great overall, I just made some minor modifications. I wonder if you could add a system test, as well as a unit test for setter with None? Please let us know if you need any help. Happy holidays!

**Comment by keunsoopark** [comment]
> system test

Hi @Linchin, thanks for the review.

I added system tests in [system/test_clients.py](https://github.com/keunsoopark/python-bigquery/blob/3d317d1b0a182df7d7f9707e243a4081dcd783de/tests/system/test_client.py#L301)

I fixed unittests for None, and added more tests for invalid values. See [test_resource_tags_setter_bad_value](https://github.com/keunsoopark/python-bigquery/blob/3d317d1b0a182df7d7f9707e243a4081dcd783de/tests/unit/test_dataset.py#L908).

Merry Christmas and happy new year!

**Comment by Linchin** [comment]
I think we also need to install the resource-manager package, similar to [this change](https://github.com/googleapis/python-bigquery/pull/2093/files#diff-f7a16a65f061822bcc73b8296f4dc837353d379d8d9cc5307982cb6941442835). Could you add this code too?

**Comment by keunsoopark** [comment]
> I think we also need to install the resource-manager package, similar to [this change](https://github.com/googleapis/python-bigquery/pull/2093/files#diff-f7a16a65f061822bcc73b8296f4dc837353d379d8d9cc5307982cb6941442835). Could you add this code too?

Thanks :) Updated!

File: tests/system/test_client.py [hunk 3]
```diff
@@ -278,33 +286,72 @@ def test_create_dataset_with_default_rounding_mode(self):
         self.assertTrue(_dataset_exists(dataset))
         self.assertEqual(dataset.default_rounding_mode, "ROUND_HALF_EVEN")
 
+    def _create_resource_tag_key_and_values(self, key, values):
```
[the remark this answers is not in the document]
Comment (Linchin) [reply tests/system/test_client.py:289]: In general you can just use pytest to run system tests, it's slightly easier than nox, and should be able to run on any python version:
```
pytest tests/system/test_client.py
```
You can specify the test class or method too:
```
pytest tests/system/test_client.py::TestBigQuery::test_update_dataset
```
This should make it faster.

[the remark this answers is not in the document]
Comment (Linchin) [reply tests/system/test_client.py:289]: I think you also need to run lint, that will need to use nox:
```
nox -r -s lint
```

File: tests/system/test_client.py [hunk 4]
```diff
@@ -278,33 +287,78 @@ def test_create_dataset_with_default_rounding_mode(self):
...
+        tag_key_client = TagKeysClient()
+        tag_value_client = TagValuesClient()
+
+        tag_key_parent = f"projects/{Config.CLIENT.project}"
+        new_tag_key = resourcemanager_types.TagKey(
+            short_name=key, parent=tag_key_parent
+        )
+        try:
+            tag_key = tag_key_client.create_tag_key(tag_key=new_tag_key).result()
+        except AlreadyExists:   # When system tests runs in parallel for multiple Python versions, the tag key may already exist
```
Comment (keunsoopark) [inline tests/system/test_client.py:300]: Since we run multiple system tests at the same time for different Python versions, we got `AlreadyExists` error for tag key only in one of two system tests.

File: tests/system/test_client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (keunsoopark) [reply tests/system/test_client.py:290]: Thanks I ran linting locally.

I still could not run system tests locally. I think I need to specify `GOOGLE_APPLICATION_CREDENTIALS` env var (see [here](https://github.com/googleapis/python-bigquery/blob/a2bebb95c5ef32ac7c7cbe19c3e7a9412cbee60d/noxfile.py#L201)), but not sure what credentials I can use.
I tried to use the credentials of a service account in my personal Google Cloud project, which is the owner of this project, but I got all `PERMISSION_DENIED` errors for dataset creation, table creation etc.

[the remark this answers is not in the document]
Comment (Linchin) [reply tests/system/test_client.py:290]: I think if you run pytest instead of nox, you don't have to have `GOOGLE_APPLICATION_CREDENTIALS` set up, because we are bypassing `noxfile.py`.

File: tests/system/test_client.py (hunk 4 above)
Comment (Linchin) [reply tests/system/test_client.py:300]: Great, thank you for fixing this!

File: tests/system/test_client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (keunsoopark) [reply tests/system/test_client.py:299]: Ok I managed running system test locally. This is how:
- Make a python script, `install_dependencies.py` to install all dependencies in `noxfile.py` in my local pip.
- For some reason, `google-cloud-testutils` was not installed in this manner, as specified [here](https://github.com/googleapis/python-bigquery/issues/1758#issuecomment-1898893183), so install this one manually.
- Configure my personal GCP project
- Run pytest

**Comment by Linchin** [comment]
The 3.12 system test is still failing - after trying to create the resource tag and finding that it already exists, the code tries to access it and gets PermissionDenied error, which indicates that the tag either doesn't exist, or the client doesn't have the permission to access it. I suspect there is a race condition where the tag was created by sys test 3.8 when 3.12 was trying to create it, and deleted when 3.12 tries to access it 😂 Let me rerun the tests to see if it's flaky.

**Comment by keunsoopark** [comment]
> The 3.12 system test is still failing - after trying to create the resource tag and finding that it already exists, the code tries to access it and gets PermissionDenied error, which indicates that the tag either doesn't exist, or the client doesn't have the permission to access it. I suspect there is a race condition where the tag was created by sys test 3.8 when 3.12 was trying to create it, and deleted when 3.12 tries to access it 😂 Let me rerun the tests to see if it's flaky, so we can verify the assumption.
> 
> If this is indeed the case, we may need to append the python version to the tags to avoid conflict.

Thanks for a good inspiration! I agree that we should make each of test running independent. Instead of postfixing python version, I postfixed random characters to tag keys. I see it works well from my local testing.

In previous run, all tests failed all of sudden, and I could no see logs any longer. Could you check if it happens again in next CI tests?

**Comment by Linchin** [comment]
Thank you for the prompt reply! Adding a prefix lgtm. I think there was something wrong with the test infra, I hope it doesn't happen again.

**Review by Linchin — APPROVED** [review]
Thank you for your contribution, now all tests have passed! 🎉

LGTM.

**Comment by keunsoopark** [comment]
> Thank you for your contribution, now all tests have passed! 🎉
> 
> LGTM.

Thank you so much! It was a pleasant review process with you!
My motivation for this contribution is to making dbt able to attach policy tags - https://github.com/dbt-labs/dbt-adapters/issues/577. As you see in this issue, it requires this functionality on tables as well.
I am looking forward to other PR - https://github.com/googleapis/python-bigquery/pull/2093 done soon, but if it takes too long time to get response, I can contribute it as well. Now I can fix this feature pretty quickly :)

**Comment by Linchin** [comment]
It was a pleasure working with you! @keunsoopark 

Let's first see if the original author of #2093 responses, then we can decide what to do next.
